File: mobile/ios/CortexV2-Emotiv-iOS/CortexV2-Emotiv-ObjC/CortexV2-Emotiv-ObjC/HeadsetFinder.m
//
//  HeadsetFinder.m
//  CortexV2-Emotiv-ObjC
//
//  Created by nvtu on 2/24/20.
//  Copyright © 2020 Emotiv. All rights reserved.
//

#import "HeadsetFinder.h"

@implementation HeadsetFinder

- (id) init {
    self = [super init];
    if (self) {
        client = [CortexClient shareInstance];
        [self handleCortexEvent];
    }
    return self;
}

-(void) handleCortexEvent {
    __unsafe_unretained typeof(self) weakSelf = self;
    client.onQueryHeadsetsOk = ^(NSArray *headsets) {
        if (headsets.count == 0) {
            return;
        }
        [weakSelf printHeadset:headsets];
        Headset *headset = headsets.firstObject;
        NSLog(@"headset status %@", headset->status);
        if ([headset->status isEqual: @"connected"]) {
            if (weakSelf.onHeadsetFound != nil) {
                [weakSelf->timer invalidate];
                weakSelf.onHeadsetFound(headset, headsets);
            }
        }
    };
}

-(void) findHeadsets {
    if (timer != nil) {
        [timer invalidate];
    }
    dispatch_async(dispatch_get_main_queue(), ^{
        self->timer = [NSTimer scheduledTimerWithTimeInterval:1 repeats:true block:^(NSTimer * timer) {
            [self->client queryHeadset:@""];
        }];
    });
}

-(void) printHeadset : (NSArray *) headsets {
    NSLog(@"%lu headset(s) found:", (unsigned long)headsets.count);
    for(Headset *hs in headsets) {
        NSLog(@"%@", [hs toString]);
    }
}
@end

File: mobile/ios/CortexV2-Emotiv-iOS/CortexV2-Example-Swift/CortexV2-Example-Swift/HeadsetFinder.swift
//
//  HeadsetFinder.swift
//  CortexV2-Example-Swift
//
//  Created by Emotiv Inc on 2/21/20.
//  Copyright © 2020 Emotiv. All rights reserved.
//

import Foundation

class HeadsetFinder {
    
    private var client = CortexClient.sharedInstance
    private var timer: Timer!
    
    var onHeadsetFound: ((Headset, [Headset]) -> Void)!
    
    init() {
        handleCortexEvent()
    }
    
    func findHeadsets() {
        if timer != nil {
            timer.invalidate()
        }
        timer = Timer.scheduledTimer(withTimeInterval: 1, repeats: true, block: { [weak self] (timer) in
            guard let weakSelf = self else { return }
            weakSelf.client.queryHeadsets()
        })
    }
    
    private func handleCortexEvent() {
        client.onQueryHeadsetsOk = { [weak self] (headsets) in
            guard let weakSelf = self else { return }
            if headsets.isEmpty {
                return
            }
            weakSelf.printHeadsets(headsets: headsets)
            let headset = headsets.first ?? Headset()
            NSLog("headset status \(headset.status)")
            if headset.status == "connected" {
                if weakSelf.onHeadsetFound != nil {
                    weakSelf.timer.invalidate()
                    weakSelf.onHeadsetFound!(headset, headsets)
                }
            }
        }
    }
    
    private func printHeadsets(headsets: [Headset])
    {
        NSLog("\(headsets.count) headset(s) found:")
        for hs in headsets {
            NSLog(hs.toString())
        }
    }
}
